bin/exec.ml

```diff
@@ -32,6 +32,104 @@ let man =
 
 let info = Cmd.info "exec" ~doc ~man
 
+module Command_to_exec = struct
+  (* A command to execute, which knows how to (re)build the program and then
+     run it with some arguments in an enivorment *)
+
+  type t =
+    { get_path_and_build_if_necessary :
+        unit -> (Path.t, [ `Already_reported ]) result Fiber.t
+    ; args : string list
+    ; env : Env.t
+    }
+
+  (* Helper function to spawn a new process running a command in an
+     environment, returning the new process' pid *)
+  let spawn_process path ~args ~env =
+    let path = Path.to_string path in
+    let env = Env.to_unix env |> Spawn.Env.of_list in
+    let argv = path :: args in
+    let pid = Spawn.spawn ~prog:path ~env ~argv () in
+    Pid.of_int pid
+
+  (* Run the command, first (re)building the program which the command is
+     invoking *)
+  let build_and_run_in_child_process
+      { get_path_and_build_if_necessary; args; env } =
+    get_path_and_build_if_necessary ()
+    |> Fiber.map ~f:(Result.map ~f:(spawn_process ~args ~env))
+end
+
+module Watch = struct
+  (* When running `dune exec` in watch mode, this will keep track of the pid of
+     the process created to run the program in the previous iteration so that
+     it can be killed (for long running programs, e.g. servers) and restarted
+     when its source is changed. *)
+
+  type state = { currently_running_pid : Pid.t option ref }
+
+  let init_state () = { currently_running_pid = ref None }
+
+  (* Send sigkill to a process and wait for it to stop. On Windows the only
+     supported (emulated) signal is sigkill. *)
+  let kill_process_windows pid =
+    let pid_int = Pid.to_int pid in
+    Unix.kill pid_int Sys.sigkill;
+    let _stopped_pid, _process_status = Unix.waitpid [] pid_int in
+    ()
+
+  (* Send sigterm to a process, and wait 1s for the process to stop. If it's
+     still running after 1s, send sigkill. *)
+  let kill_process_unix pid =
+    let pid_int = Pid.to_int pid in
+    Unix.kill pid_int Sys.sigterm;
+    let period = 0.1 in
+    (* Periodically check if the process is still running ([wait]-ing with
+       [WNOHANG]) after sending sigterm, and send sigkill if its still running
+       after a timeout *)
+    let rec poll remaining_time =
+      if remaining_time <= 0.0 then (
+        Unix.kill pid_int Sys.sigkill;
+        let _stopped_pid, _process_status = Unix.waitpid [] pid_int in
```

Comment:
Note that `waitpid` on a process isn't allowed inside a Fiber. The scheduler already calls `Unix.waitpid [] -1` periodically, so it might reap your process already. So if you really want your own waitpid loop, it should run outside the fiber.

Reply:
ah I didn't realize that. I've updated it to use `Scheduler.wait_for_process` rather than directly calling `waitpid`.